### ISSUE TITLE
stand alone library?

### ISSUE THREAD

**Issue by byersolomongmail** [curlconverter/curlconverter#648]
it would be very useful to have the stand-alone library hosted on jsdeliver with the tree-sitter and everything it is very hard to follow the node_modules packaging and the massive tree-sitter.

**Comment by verhovsky** [comment]
> have the stand-alone library hosted on jsdeliver

I looked at their website and I don't see what I have to do to get them to host it. Don't they just host everything automatically?

**Comment by byersolomongmail** [comment]
IDK it's there and doesn't work gives errors. I am not as good at node and stuff. https://www.jsdelivr.com/package/npm/curlconverter-release

**Comment by chris48s** [comment]
So I tried to do this a few days ago. This was my experience trying to use curlconverter from https://www.jsdelivr.com/package/npm/curlconverter

---

```html
<script src="https://cdn.jsdelivr.net/npm/curlconverter@4.9.0/dist/src/index.min.js"></script>
```

throws

```
Uncaught SyntaxError: export declarations may only appear at top level of a module
```

because it is an ESModule

---

```html
<script type="module" src="https://cdn.jsdelivr.net/npm/curlconverter@4.9.0/dist/src/index.min.js"></script>
```

throws

```
Uncaught TypeError: The specifier “yamljs” was a bare specifier, but was not remapped to anything. Relative module specifiers must start with “./”, “../” or “/”.
```

---

So then you have to build an importmap to resolve each dependency from jsDelivr:

```html
<script type="importmap">
  {
    "imports": {
      "yamljs": "https://cdn.jsdelivr.net/npm/yamljs@0.3.0/dist/yaml.min.js",
      "tree-sitter": "https://cdn.jsdelivr.net/npm/tree-sitter@0.21.1/index.min.js",
      // ..and so on
    }
  }
</script>
<script type="module" src="https://cdn.jsdelivr.net/npm/curlconverter@4.9.0/dist/src/index.min.js"></script>
```

at this point I gave up, fired up webpack, and just built my own bundle :D

---

I think the request here is to publish a built asset as part of the package which has all the dependencies bundled so that it can be pulled in as a single file from jsDelivr that is usable in the browser.

**Comment by verhovsky** [comment]
I don't think it's possible to use libraries that use Wasm (which curlconverter does) through jsDelivr

**Comment by byersolomongmail** [comment]
Well maybe just the import map somewhere would be good enough so the user doesn't have to find it all.

**Comment by dosentmatter** [comment]
You can actually download the `wasm` binaries from the CDNs:

https://cdn.jsdelivr.net/npm/curlconverter/dist/tree-sitter-bash.wasm
https://cdn.jsdelivr.net/npm/web-tree-sitter/tree-sitter.wasm

https://esm.sh/web-tree-sitter/tree-sitter.wasm
https://esm.sh/curlconverter/dist/tree-sitter-bash.wasm

One issue is that `curlconverter` hardcodes the paths to be [`/tree-sitter-bash.wasm` and `/tree-sitter.wasm`](https://github.com/curlconverter/curlconverter/blob/v4.12.0/src/shell/webParser.ts#L7-L13). When fetched from a CDN, I'm not if these paths are relative to the CDN's origin or the current website's origin.

I tried to proxy
https://esm.sh/tree-sitter.wasm -> https://esm.sh/web-tree-sitter/tree-sitter.wasm
https://esm.sh/tree-sitter-bash.wasm -> https://esm.sh/curlconverter/dist/tree-sitter-bash.wasm

But it still failed with `fs` errors. For some reason https://esm.sh/curlconverter and https://esm.sh/web-tree-sitter have imports of https://esm.sh/node/buffer.mjs and https://esm.sh/node/process.mjs which causes `process` to be defined and `ENVIRONMENT_IS_NODE === true`. I tried to get rid of the `fs` using [`esm.sh` query parameters](https://esm.sh/#docs) but wasn't successful. `web-tree-sitter` says to use `resolve.fallback` in webpack config, but I'm not sure if `esm.sh` has an equivalent query parameter https://www.npmjs.com/package/web-tree-sitter/v/0.25.3#user-content-cant-resolve-fs-in-node_modulesweb-tree-sitter

![Image](https://github.com/user-attachments/assets/bf664a67-2b6d-48cb-a976-6963a5ba7e60)

![Image](https://github.com/user-attachments/assets/90696147-f0bf-4b7b-9b3f-166980f82cdd)

![Image](https://github.com/user-attachments/assets/479aec5b-ce80-44fe-8a7d-9803e50e2cd0)

If I forcefully set `ENVIRONMENT_IS_NODE = false` in the debugger, I get some other wasm error:

![Image](https://github.com/user-attachments/assets/74ba6661-38b2-42bb-b65f-9d0017e5070a)

I also noticed that `curlconverter@3` didn't use `tree-sitter` so I tried using that in the browser, but it failed because it uses a fork of `yargs`, `@curlconverter/yargs`, which seems to be built for nodejs.

**Comment by verhovsky** [comment]
> which seems to be built for nodejs.

The goal was to make a single npm module that works seamlessly on both nodejs and the browser

**Comment by dosentmatter** [comment]
So it seems like esm.sh is bundling the node dependencies because modules like `fs`, `buffer`, `process`, and `path` are in the source code, even though they aren't used in the browser based on condtions like in

https://github.com/curlconverter/curlconverter/blob/8263e92e292764f7ef25ea7681fb08db9f51cfeb/src/utils.ts#L16-L20

I found that https://esm.sh/curlconverter?external=node:fs,node:path,node:process,node:buffer works to externalize the modules and turn the imports from `'/node/process'` into `'node:process'`.

I tried to alias these node imports into noops with https://esm.sh/curlconverter?alias=node:process:noop3, but it had no effect. I think it might be due to conflicts with the `:` delimiter syntax. I don't see any query parameter in esm.sh docs that would allow bundling these node modules as globals that don't get used in the browser.

So then instead, I made a simple HTML file with `importmap`. Initially, I also had `node:buffer` in the `importmap`, but then it was complaining because `noop3` doesn't have  `import { Buffer } from 'node:buffer'`. You can see I commented this out. Then I changed it to just having `node:process`. I want to make `process === undefined`, so it won't set `ENVIRONMENT_IS_NODE === true`.

```html
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8" />
    <title>Import Map Alias for node: Modules</title>
    <script type="importmap">
      {
        "imports": {
          "node:process": "https://esm.sh/noop3"
        }
      }
    </script>
    <!--
    <script type="importmap">
      {
        "imports": {
          "node:fs": "https://esm.sh/noop3",
          "node:path": "https://esm.sh/noop3",
          "node:process": "https://esm.sh/noop3",
          "node:buffer": "https://esm.sh/noop3"
        }
      }
    </script>
    -->
</head>

<body>
    <script type="module">
        // import { convert } from "https://esm.sh/curlconverter?external=node:fs,node:path,node:process,node:buffer";
        import * as curlconverter from "https://esm.sh/curlconverter?external=node:process";
        console.log(curlconverter);
    </script>
</body>

</html>
```

I was opening the HTML from a file so it had CORS error when trying to fetch the wasm. It looks like it tries to fetch the wasm from the current page's origin and not from the CDN's origin. So my proxy didn't have any effect anymore. Previously, I was testing in the browser dev console while on https://esm.sh site, so it would fetch from esm.sh origin
https://esm.sh/tree-sitter.wasm -> https://esm.sh/web-tree-sitter/tree-sitter.wasm
https://esm.sh/tree-sitter-bash.wasm -> https://esm.sh/curlconverter/dist/tree-sitter-bash.wasm
<img width="1728" alt="Image" src="https://github.com/user-attachments/assets/e179efe8-4e64-41e6-b40c-f9cf5afe9e86" />

So then I served up the HTML with a local http static server. And added the wasm in the directory. Then it has similar wasm error as when I forcefully set `ENVIRONMENT_IS_NODE = false`
<img width="1727" alt="Image" src="https://github.com/user-attachments/assets/278c22af-039f-4107-92d7-710ee0990140" />
